File: Strings/longest_substring.py
def longest_substring_without_duplication(string):
    longest = [0, 1]
    start = 0
    dict = {}

    for i, char in enumerate(string):
        if char in dict:
            start = max(start, dict[char] + 1)
        if longest[1] - longest[0] < i + 1 - start:
            longest = [start, i + 1]
        dict[char] = i
    return string[longest[0] : longest[1]]

File: Strings/longestSubString.js
const longestSubstringWithoutDuplication = (string) => {
    let longestStr = [0, 1];
    let start = 0;
    let dict = {};

    for (let i = 0; i < string.length; i++) {
        const char = string[i];
        if (char in dict) {
            start = Math.max(start, dict[char] + 1);
        }
        if (longestStr[1] - longestStr[0] < i + 1 - start) {
            longestStr = [start, i + 1];
        }
        dict[char] = i;
    }
    return string.slice(longestStr[0], longestStr[1]);
}